### PR TITLE
feat(verify): detect test tampering

### LINKED ISSUE #1072: feat(verify): detect reward hacking & test tampering
## Motivation

Reward hacking is the most heavily-evidenced and *least-defended* agent failure mode in Sybra today. METR found o3 reward-hacked 100% of one task; EvilGenie (Nov 2025) saw 42–77% of agents hardcode test cases and Gemini CLI delete the test file outright; Anthropic showed gaming *generalizes to sabotage* (12% sabotage attempts after RL on hack-vulnerable envs). OpenAI retired SWE-bench Verified after agents mined git history for the fix.

Sybra has zero defense. `verify` today (`internal/workflow/engine_steps_verify.go:execVerifyCommits`) only counts commits ahead of `origin/main` — it cannot tell whether the agent passed by *solving* or by *gaming*. This complements #1058 (which runs the project's checks); this issue is about detecting gaming **of** those checks.

## Implementation information

- Diff inspector: flag implementation diffs that touch test files, snapshots, fixtures, or CI config. Classify suspicious edits — added `skip`/`xfail`/`exit(0)`/`return <expected>`, deleted assertions, hardcoded expected values.
- Run the **baseline suite from `origin/main` (Pass-to-Pass)** AND the **agent's suite (Fail-to-Pass)**; both must pass. The agent never saw the baseline run, so it cannot pre-game it. This is the SWE-bench verification model and the cheapest robust anti-gaming check.
- Surface flags to the adversarial reviewer / `human-required`; for autonomous tasks, any test-file change requires an explicit bless before `in-review → done`.
- Prompt floor: add "never weaken, skip, delete, or hardcode tests to make them pass" to implementation + PR-fix prompts (`internal/sybra/app_reviews_fix.go`, impl templates).

## Acceptance

A task that deletes/weakens a test or hardcodes an expected value is detected, flagged, and blocked from reaching `done` without human bless.

## Dependencies

Depends on #1058 landing the check-runner harness (baseline-suite execution). Detector layer can be built alongside.

### PR DESCRIPTION
**Pull request by Automaat**
## Motivation

Reward hacking is Sybra's least-defended agent failure mode. `verify`
today only counts commits ahead of `origin/main` — it cannot tell whether
an agent passed by solving the task or by gaming the checks (deleting a
test, adding `t.Skip`, commenting out an assertion, neutering CI). This
adds the mechanical diff-inspector layer of #1072 so a gamed change is
detected and cannot reach `done` without an explicit human bless.

> Changelog: feat(verify): detect test tampering

## Implementation information

New `detect_tampering` workflow step (`internal/workflow/engine_steps_tamper.go`):

- Runs `git diff --name-status base...HEAD`, classifies changed files
  (test / snapshot / fixture / CI / other), and scans the per-file
  patches of verification files for high-severity signals:
  added skip/xfail/`exit(0)`/`//go:build ignore`; commented-out or
  net-removed assertions and test cases; tautological self-comparisons
  (`Equal(t, got, got)`); deleted verification files; neutered CI gates
  (`continue-on-error: true`, `|| true`, `exit 0`).
- A high finding flips the task to `human-required` with the offending
  rules in the status reason; benign test edits are recorded in a
  `tamper-report.json` artifact (via the new `ArtifactRecorder.PutGeneric`)
  but do not block.
- Hunk-aware diff parsing (a content line beginning with `--`/`++` is not
  mistaken for a file header) and `core.quotePath=false` for unicode paths.
- A `tamper-blessed` tag short-circuits the scan so a human-accepted diff
  is not re-flagged on re-dispatch (no livelock). To accept a flagged
  change: add the `tamper-blessed` tag, then move the task back into flow.

Wired into the three paths where reward-hacking pressure exists —
`simple-task-implement` (after `verify_commits`), `simple-task-review`
(after `fix_review`), and `pr-fix` (after `verify_commits`). Each routes a
flagged task to `human-required` and ends the workflow before a PR opens.

Prompt floor added to the implementation and PR/review-fix prompts:
never weaken, skip, delete, comment out, or hardcode tests.

### Scope / follow-ups

- This is the diff-inspector layer only. Pure literal-value hardcoding
  (an expected `5` swapped to `4`) is undetectable from the diff alone;
  robust coverage comes from the **baseline-suite (Pass-to-Pass)**
  execution, which depends on the check-runner harness in #1058 and lands
  as a follow-up. The tautology check covers the common lazy case.
- Net-counting favors detection over false-negatives, so some legitimate
  refactors (moving assertions into a helper) may flag — the
  `tamper-blessed` escape hatch makes that a one-tag action.

## Supporting documentation

Closes #1072. Depends conceptually on #1058 for the deferred baseline-suite layer.

Table-driven unit tests cover path classification, every detection rule,
diff parsing edge cases, the bless escape, and the workflow wiring on all
three paths.

### CODE REVIEW

File: internal/workflow/engine_steps_tamper.go [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+	}
+
+	report, err := e.collectTamperReport(taskID, wtPath)
+	if err != nil {
+		if errors.Is(e.ctx.Err(), context.Canceled) || errors.Is(e.ctx.Err(), context.DeadlineExceeded) {
+			return StepOutput{StepID: step.ID, Status: "completed", Output: "skipped: context canceled"}, nil
+		}
+		e.logger.Warn("workflow.detect-tampering.diff-error", "task_id", taskID, "err", err)
+		return StepOutput{StepID: step.ID, Status: "completed", Output: "clean"}, nil
+	}
```
[the remark this answers is not in the document]
Comment (Automaat) [reply internal/workflow/engine_steps_tamper.go:442]: Fixed in 0191bc5. collectTamperReport now wraps the per-step ctx.Err() into the returned error, and execDetectTampering checks the error chain (errors.Is(err, context.DeadlineExceeded/Canceled)) instead of e.ctx, so a shellTimeout is reported as canceled rather than silently clean.

File: internal/workflow/engine_steps_tamper.go [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+		if classifyTamperPath(c.Path) == tamperCatOther || strings.HasPrefix(c.Status, "D") {
+			continue
+		}
+		patch, pErr := gitFilePatch(ctx, wtPath, rangeSpec, c.Path)
+		if pErr != nil {
+			e.logger.Warn("workflow.detect-tampering.file-patch", "task_id", taskID, "file", c.Path, "err", pErr)
+			continue
+		}
+		c.Patch = patch
+	}
```
[the remark this answers is not in the document]
Comment (Automaat) [reply internal/workflow/engine_steps_tamper.go:514]: Fixed in 0191bc5. collectTamperReport now stops fetching per-file patches once it has fetched tamperMaxScannedFiles verification files, matching the cap buildTamperReport scans under, so a large diff can't spawn unbounded git diff calls.

File: internal/workflow/engine_steps_tamper.go [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+		`\bos\.Exit\s*\(\s*0\s*\)|\bsys\.exit\s*\(\s*0\s*\)|\bprocess\.exit\s*\(\s*0\s*\)|(^|[^.\w])exit\s*\(\s*0\s*\)`)
+	// tamperBuildIgnoreRe matches an added Go build-ignore tag (excludes the
+	// file from the build, so its tests stop running).
+	tamperBuildIgnoreRe = regexp.MustCompile(`^//go:build\s+ignore\b|^// \+build\s+ignore\b`)
+	// tamperAssertionRe matches an assertion line. Counted on both sides of the
+	// diff so a net removal (deletions > additions) flags weakened coverage
+	// without false-positiving on pure refactors/renames.
+	tamperAssertionRe = regexp.MustCompile(
+		`\bassert\b|\brequire\b|\bassert\.|\brequire\.|\bexpect\s*\(|` +
+			`\bt\.(Error|Errorf|Fatal|Fatalf)\b|\bEXPECT_|\bASSERT_|\.should\b|\bshould\.`)
```
[the remark this answers is not in the document]
Comment (Automaat) [reply internal/workflow/engine_steps_tamper.go:143]: Fixed in 0191bc5. Dropped bare `require` from tamperAssertionRe; only `require.` (testify-style) counts now, so removing a JS `require("dep")` / Ruby `require "spec_helper"` import no longer triggers removed-assertions. Added a regression test (js_require_import_removal_is_not_an_assertion).

File: internal/workflow/engine_steps_tamper.go [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+			content := line[1:]
+			if isCI || looksLikeComment(content) {
+				continue
+			}
+			if tamperAssertionRe.MatchString(content) {
+				delAssert++
+			}
+			if tamperTestDeclRe.MatchString(content) {
+				delDecl++
+			}
```
[the remark this answers is not in the document]
Comment (Automaat) [reply internal/workflow/engine_steps_tamper.go]: Fixed in 0191bc5. Added tamperCIRunRe and net-counting of check-running steps in CI files: a net removal of `go test`/lint/etc. run steps now flags removed-ci-step (high), so deleting a gate without a neuter token is caught. Tests: ci_removed_test_step, ci_renamed_step_no_net_loss.
